docs/devGuide/migratingToTypeScript.md

```diff
@@ -0,0 +1,131 @@
+
+### Step 1: Preparation
+
+1. **Install typings of external dependencies** relevant to the files, if any.
+
+   Generally, external modules have their typings published as another module named `@types/<lib-name>` (e.g. `@types/jest`) or bundled alongside their implementation. You can check which is the case [here](https://www.typescriptlang.org/dt/search?search=). Check for `@types/<lib-name>` first, then check for `<lib-name>` itself if not found.
+
+   If the dependency has its typings bundled alongside their implementation, you do not need to do anything else.
+
+   If the dependency has a `@types` module, install the typings as a development dependency for the `core` package by:
+
+   ```bash {.no-line-numbers}
+   npx lerna add --dev @types/<lib-name> --scope @markbind/core
+   ```
+
+   Note that you can only install one package at a time with `lerna add`. Verify the addition in `packages/core/package.json`.
+
+   <box type="info">
+
+   It's recommended to try to match the `@types/<lib-name>` version with the associated module. In general, try to match the major and minor version as closely as possible.
+   </box>
+
+2. **Keep the changes** (`packages/core/package.json` and root `package-lock.json`) to be commited in Step 3 later.
+
+### Step 2: "Rename" Commit
+
+1. **Disable automatic compilation of TypeScript files** on save/change.
+
+   You can read more about this on [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.
+
+2. **Add the paths to the `.js` files to `.gitignore` and `.eslintignore`**. This will let Git and ESLint ignore the JavaScript files compiled by TypeScript later.
+
+3. **Rename the files** from having `.js` extensions to `.ts` extensions.
+
+4. **Stage the changes** (the renamed files, `.gitignore`, and `.eslintignore`) for commit. Ensure the renamed files are regarded by Git as `renamed` and not `added`.
+
+5. **Commit the staged changes** with the message `Rename <file-or-folder> to TypeScript`.
+   <br>
+   For example: `Rename core/src/html to TypeScript`.
+
+### Step 3: "Adapt" Commit
+
+1. **Enable automatic compilation of TypeScript files** on save/change.
+
+   You can read more about this on [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.
+
+2. **Change the CommonJS module import/export syntax** to TypeScript's equivalent (use ES6 only if necessary).
+
+   A common error in newly renamed files comes from the way modules are imported/exported. The syntax for those can be converted according to the table below:
+
+   CommonJS                | TypeScript equivalent    | ES6
+   ------------------------|--------------------------|-----------------------
+   `const ... = require()` | `import ... = require()` | `import ... from ...`
+   `module.exports = {}`   | `export = {}`            | `export ...`
+   `module.exports = ...`  | `export = ...`           | `export default ...`
```

Comment:
> Do not use the ES6 syntax if not necessary, as it compiles into something more than just the CommonJS syntax when compiled, and might lead to unexpected behaviours when another `.js` file imports the compiled file, to the point of failing established tests in some cases.

This particular line may not be equivalent (commonJs dosen't have concept of `default`), and might be the sole cause of your issues with above quote.

Try inserting the following in the playground [here](https://www.typescriptlang.org/play?module=1#code/PTAEAEGcAsHsHcCiBbAlgFwFAgs2ATAVwBsBTALlAGNZk8A7AK0mzHHtkQCcvYuXUyAA590oAN6gAbgENihUgHkAZgAVUoAL6hlvZKABEAOmA16kdDPrpIBgNyZMpAB4iuYsxdDp4sdaABeaTkFFX8AKlAAJgcnV1FQfFJlGRIxHz9UByA)

```
// @showEmit
// @module: commonjs
// @noErrors
import { valueOfPi } from "./constants";

export const twoPi = valueOfPi * 2;

export default twoPi;
```

Reply:
I think one solution may be to advise against using `export default` in our **intra** (external libraries should be fine) exports / imports altogether (we've been mostly using named exports too which should make this easier)

Reply:
Agree with `export default`. I have already disabled `prefer-default-export` in ESLint, maybe we can strengthen it by enabling `no-default-export`. Does external libraries here mean the ones that will be consumed by other packages (e.g. `cli`)? If so, we still have to follow how `cli/index.js` imports `core/index.js` to see our options right now.

As a note, I was just about to write that I remembered I couldn't make ES6-style named exports (`export ...`) work in `CyclicReferenceError.ts` as it will always fail the tests, suggesting that it's not always possible to use ES6-style. But, as I investigated further to write this comment, I found that the root cause is that there is a file (`Context.js`) that imports the `CyclicReferenceError` class as a *default* export. Because `export default` is out of the question, therefore the only way for `CyclicReferenceError.ts` to fit in with other files is to use the TypeScript-style default export (`export = ...`).

I can update the text so that we can start from TypeScript-style, then move to ES6 *if possible*, but do not use `export default`.

Reply:
> Does external libraries here mean the ones that will be consumed by other packages (e.g. cli)?

Nope, just regular external libraries (e.g. `simple-git`).

> I can update the text so that we can start from TypeScript-style, then move to ES6 if possible, but do not use export default.

Sounds good 👍
(at least for now, while we're still migrating, default exports won't play so well with js)

> I found that the root cause is that there is a file (Context.js) that imports the CyclicReferenceError class as a default export.

> default export.

Rather, the `module.exports` object. (commonjs has no "default" export, by default 😅)

Just to be clear, this would actually work, but rather clunky-ly:

```
var x = require('ts-module').default;
```

Once we have almost all files converted we can possibly revisit if `export default` would be a good fit (if really needed).